dotnet/src/Connectors/Connectors.HuggingFace/Core/HuggingFaceClient.cs

```diff
@@ -12,58 +12,134 @@
-        this._separator = endpoint.AbsolutePath.EndsWith("/", StringComparison.InvariantCulture) ? string.Empty : "/";
-        this._endpoint = endpoint;
-        this._modelId = modelId;
-        this._apiKey = apiKey;
+        this.Separator = endpoint.AbsolutePath.EndsWith("/", StringComparison.InvariantCulture) ? string.Empty : "/";
+        this.Endpoint = endpoint;
+        this.ModelId = modelId;
+        this.ApiKey = apiKey;
         this._httpClient = httpClient;
-        this._logger = logger ?? NullLogger.Instance;
-        this._streamJsonParser = streamJsonParser ?? new StreamJsonParser();
+        this.Logger = logger ?? NullLogger.Instance;
+
+        this._openAIClient = new HuggingFaceOpenAIClient(this);
+    }
+
+    #region ClientCore
+    internal static void ValidateMaxTokens(int? maxTokens)
+    {
+        if (maxTokens is < 1)
+        {
+            throw new ArgumentException($"MaxTokens {maxTokens} is not valid, the value must be greater than zero");
+        }
+    }
+
+    internal async Task<string> SendRequestAndGetStringBodyAsync(
+        HttpRequestMessage httpRequestMessage,
+        CancellationToken cancellationToken)
+    {
+        using var response = await this._httpClient.SendWithSuccessCheckAsync(httpRequestMessage, cancellationToken)
+            .ConfigureAwait(false);
+
+        var body = await response.Content.ReadAsStringWithExceptionMappingAsync()
+            .ConfigureAwait(false);
+
+        return body;
+    }
+
+    internal async Task<HttpResponseMessage> SendRequestAndGetResponseImmediatelyAfterHeadersReadAsync(
+        HttpRequestMessage httpRequestMessage,
+        CancellationToken cancellationToken)
+    {
+        var response = await this._httpClient.SendWithSuccessCheckAsync(httpRequestMessage, HttpCompletionOption.ResponseHeadersRead, cancellationToken)
+            .ConfigureAwait(false);
+        return response;
+    }
+
+    internal static T DeserializeResponse<T>(string body)
+    {
+        try
+        {
+            T? deserializedResponse = JsonSerializer.Deserialize<T>(body);
+            if (deserializedResponse is null)
+            {
+                throw new JsonException("Response is null");
+            }
+
+            return deserializedResponse;
+        }
+        catch (JsonException exc)
+        {
+            throw new KernelException("Unexpected response from model", exc)
+            {
+                Data = { { "ResponseData", body } },
+            };
+        }
+    }
+
+    internal void SetRequestHeaders(HttpRequestMessage request)
+    {
+        request.Headers.Add("User-Agent", HttpHeaderConstant.Values.UserAgent);
+        request.Headers.Add(HttpHeaderConstant.Names.SemanticKernelVersion, HttpHeaderConstant.Values.GetAssemblyVersion(this.GetType()));
+        if (!string.IsNullOrEmpty(this.ApiKey))
+        {
+            request.Headers.Add("Authorization", $"Bearer {this.ApiKey}");
+        }
```

Comment:
this apiKey/bearerToken has not expiration? Maybe introduce same `Func<Task>` like in gemini.

Reply:
No, in hugging face the concept is an ApiKey which is long lasting.